File: pi-agent/extensions/pi-inline-images/test/private-fixture/build-session.mjs
import { createHash } from "node:crypto";
import { mkdirSync, writeFileSync } from "node:fs";
import { resolve } from "node:path";
import sharp from "sharp";

const work = process.argv[2];
if (!work) throw new Error("usage: node build-session.mjs WORK_DIR");
mkdirSync(resolve(work, "images"), { recursive: true });
mkdirSync(resolve(work, "sessions"), { recursive: true });
mkdirSync(resolve(work, "agent"), { recursive: true });
const sha256 = (value) => createHash("sha256").update(value).digest("hex");
const entries = [];
let parentId = null;
let sequence = 0;
const append = (entry) => {
  const id = `fixture${String(++sequence).padStart(4, "0")}`;
  entries.push({ ...entry, id, parentId, timestamp: "2026-09-16T00:00:00.000Z" });
  parentId = id;
};

const previews = [];
for (let index = 0; index < 20; index++) {
  const width = 80, height = 8;
  const raw = Buffer.alloc(width * height * 4);
  for (let pixel = 0; pixel < width * height; pixel++) {
    raw[pixel * 4] = (index * 41 + pixel * 3) & 0xff;
    raw[pixel * 4 + 1] = (index * 83 + pixel * 5) & 0xff;
    raw[pixel * 4 + 2] = (index * 127 + pixel * 7) & 0xff;
    raw[pixel * 4 + 3] = 255;
  }
  const png = await sharp(raw, { raw: { width, height, channels: 4 } }).png({ compressionLevel: 0 }).toBuffer();
  const data = png.toString("base64");
  const contentHash = sha256(data);
  const rawHash = sha256(png);
  const toolCallId = `fixture-read-${String(index).padStart(2, "0")}`;
  append({
    type: "message",
    message: {
      role: "toolResult",
      toolCallId,
      toolName: "read",
      content: [{ type: "text", text: `fixture read ${index}` }, { type: "image", mimeType: "image/png", data }],
      isError: false,
      timestamp: 0,
    },
  });
  previews.push({
    path: "attached image",
    hash: rawHash,
    originalMime: "image/png",
    width,
    height,
    logicalId: `fixturepreview${String(index).padStart(4, "0")}`,
    origin: { messageOrdinal: index, key: `tool:${toolCallId}`, blockIndex: 1, mimeType: "image/png", contentHash },
    ...(index === 19 ? {
      readProvenance: {
        version: 1,
        status: "unverified",
        toolCallId,
        blockIndex: 1,
        blockHash: contentHash,
        receivedHash: rawHash,
        receivedMime: "image/png",
        reason: "tool-source-unverified",
      },
    } : {}),
  });
}

const width = 1920, height = 1080;
const largeRaw = Buffer.allocUnsafe(width * height * 4);
let state = 0x51f15e5d;
for (let index = 0; index < largeRaw.length; index++) {
  state = (Math.imul(state, 1664525) + 1013904223) >>> 0;
  largeRaw[index] = state >>> 24;
}
const largePng = await sharp(largeRaw, { raw: { width, height, channels: 4 } }).png({ compressionLevel: 0 }).toBuffer();
const largePath = resolve(work, "images", "large-1920x1080.png");
writeFileSync(largePath, largePng);
for (const preview of previews) append({ type: "custom", customType: "pi-tmux-images.preview", data: preview });
append({
  type: "message",
  message: {
    role: "assistant",
    content: [{ type: "text", text: `NATIVE-MARKDOWN-BEGIN\n\n![1920x1080 alpha fixture](${largePath})\n\nNATIVE-MARKDOWN-END` }],
    api: "fixture",
    provider: "none",
    model: "none",
    usage: { input: 0, output: 0, cacheRead: 0, cacheWrite: 0, totalTokens: 0, cost: { input: 0, output: 0, cacheRead: 0, cacheWrite: 0, total: 0 } },
    stopReason: "stop",
    timestamp: 0,
  },
});

const session = [
  { type: "session", version: 3, id: "55555555-5555-4555-8555-555555555555", timestamp: "2026-09-16T00:00:00.000Z", cwd: work },
  ...entries,
];
const sessionPath = resolve(work, "sessions", "fixture.jsonl");
writeFileSync(sessionPath, `${session.map((entry) => JSON.stringify(entry)).join("\n")}\n`);
writeFileSync(resolve(work, "agent", "settings.json"), "{}\n");
writeFileSync(resolve(work, "agent", "models-store.json"), "{}\n");
writeFileSync(resolve(work, "manifest.json"), `${JSON.stringify({
  sessionPath,
  largePath,
  largeBytes: largePng.length,
  largeSha256: sha256(largePng),
  largePixelSha256: sha256(largeRaw),
  expectedUploadsPerViewer: 17,
  previewEntries: 20,
  retainedReadEntries: 16,
}, null, 2)}\n`);
console.log(sessionPath);

File: pi-agent/extensions/pi-inline-images/test/private-fixture/duplicate-read/build-session.py
#!/usr/bin/env python3
import base64, hashlib, json, sys
from pathlib import Path
from PIL import Image, ImageDraw

work = Path(sys.argv[1]).resolve()
(work / "images").mkdir(parents=True, exist_ok=True)
(work / "sessions").mkdir(parents=True, exist_ok=True)
(work / "agent").mkdir(parents=True, exist_ok=True)
width, height = 800, 300
image = Image.new("RGB", (width, height), (15, 18, 28))
d = ImageDraw.Draw(image)
d.rectangle((0, 0, width - 1, height - 1), fill=(255, 0, 180))
d.rectangle((18, 18, width - 19, height - 19), fill=(0, 220, 255))
d.rectangle((400, 18, width - 19, height - 19), fill=(255, 230, 0))
d.rectangle((55, 65, 350, 235), fill=(30, 60, 220))
d.rectangle((450, 65, 745, 235), fill=(30, 220, 80))
d.ellipse((325, 55, 475, 205), fill=(255, 50, 35), outline=(255, 255, 255), width=12)
source = work / "images" / "duplicate-source.png"
image.save(source, format="PNG", compress_level=0)
raw = source.read_bytes(); data = base64.b64encode(raw).decode("ascii")
raw_hash = hashlib.sha256(raw).hexdigest(); content_hash = hashlib.sha256(data.encode()).hexdigest()
calls = ["duplicate-read-call-0001", "duplicate-read-call-0002"]
usage = {"input": 0, "output": 0, "cacheRead": 0, "cacheWrite": 0, "totalTokens": 0,
         "cost": {"input": 0, "output": 0, "cacheRead": 0, "cacheWrite": 0, "total": 0}}
entries = []; parent = None; sequence = 0
def append(entry):
    global parent, sequence
    sequence += 1; ident = f"duplicate{sequence:04d}"
    entries.append({**entry, "id": ident, "parentId": parent, "timestamp": "2026-09-16T00:00:00.000Z"}); parent = ident
append({"type": "message", "message": {"role": "assistant", "content": [
    {"type": "toolCall", "id": call, "name": "read", "arguments": {"path": str(source)}} for call in calls
], "api": "fixture", "provider": "none", "model": "none", "usage": usage, "stopReason": "toolUse", "timestamp": 0}})
for call in calls:
    append({"type": "message", "message": {"role": "toolResult", "toolCallId": call, "toolName": "read",
        "content": [{"type": "text", "text": "Read image file [image/png]"}, {"type": "image", "mimeType": "image/png", "data": data}],
        "isError": False, "timestamp": 0}})
for index, call in enumerate(calls):
    preview = {"path": "attached image", "hash": raw_hash, "originalMime": "image/png", "width": width, "height": height,
        "logicalId": f"duplicatepreview{index:016d}",
        "origin": {"messageOrdinal": index + 1, "key": f"tool:{call}", "blockIndex": 1, "mimeType": "image/png", "contentHash": content_hash},
        "readProvenance": {"version": 1, "status": "unverified", "toolCallId": call, "blockIndex": 1,
            "blockHash": content_hash, "receivedHash": raw_hash, "receivedMime": "image/png", "reason": "tool-source-unverified"}}
    append({"type": "custom", "customType": "pi-tmux-images.preview", "data": preview})
append({"type": "message", "message": {"role": "assistant", "content": [{"type": "text", "text": "DUPLICATE-READ-END"}],
    "api": "fixture", "provider": "none", "model": "none", "usage": usage, "stopReason": "stop", "timestamp": 0}})
session = [{"type": "session", "version": 3, "id": "66666666-6666-4666-8666-666666666666",
            "timestamp": "2026-09-16T00:00:00.000Z", "cwd": str(work)}, *entries]
session_path = work / "sessions" / "duplicate.jsonl"
session_path.write_text("\n".join(json.dumps(x, separators=(",", ":")) for x in session) + "\n")
(work / "agent" / "settings.json").write_text("{}\n"); (work / "agent" / "models-store.json").write_text("{}\n")
(work / "manifest.json").write_text(json.dumps({"sessionPath": str(session_path), "sourcePath": str(source), "toolCallIds": calls,
    "blockIndex": 1, "expectedVisibleOccurrences": len(calls), "bytes": len(raw), "sha256": raw_hash,
    "contentHash": content_hash, "width": width, "height": height}, indent=2) + "\n")
print(session_path)
